### PR TITLE
Fix for race conditons while updating avi model with for EVH

### PR DESCRIPTION
**Pull request by monotosh-avi**
In the rest layer we take lock and get a copy of the model.
But in layer2 we were not always taking a lock before updating the models.

A similar fix was done for shard vs model already.

### CODE REVIEW

**Comment by monotosh-avi** [comment]
Similar fix with shard vs model
https://github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/pull/408

**Comment by monotosh-avi** [comment]
build ako

**Comment by monotosh-avi** [comment]
build ako

**Comment by monotosh-avi** [comment]
build ako

**Review by akshayhavile — COMMENTED** [review]
@monotosh-avi: Similar changes will be required for insecure ingress function: ProcessInsecureHostsForEVH()

**Review by chauhanshubham — DISMISSED** [review]

File: internal/nodes/avi_model_evh_nodes.go [hunk 1]
```diff
@@ -989,103 +989,112 @@ func evhNodeHostName(routeIgrObj RouteIngressModel, tlssetting TlsSettings, ingN
...
-		certsBuilt := false
-		evhSecretName := tlssetting.SecretName
-		re := regexp.MustCompile(fmt.Sprintf(`^%s.*`, lib.DummySecret))
-		if re.MatchString(evhSecretName) {
-			certsBuilt = true
+		modelGraph := aviModel.(*AviObjectGraph)
+		modelGraph.BuildModelGraphForSecureEVH(routeIgrObj, ingressHostMap, hosts, tlssetting, ingName, namespace, infraSettingName, host, key, paths)
+		// Only add this node to the list of models if the checksum has changed.
+		utils.AviLog.Debugf("key: %s, Saving Model: %v", key, utils.Stringify(vsNode))
+		modelChanged := saveAviModel(model_name, aviModel.(*AviObjectGraph), key)
```
[the remark this answers is not in the document]
Comment (monotosh-avi) [reply internal/nodes/avi_model_evh_nodes.go:996]: Done. Took care of the changes in ProcessInsecureHostsForEVH().

File: internal/nodes/avi_model_evh_nodes.go [hunk 2]
```diff
@@ -740,77 +740,83 @@ func ProcessInsecureHostsForEVH(routeIgrObj RouteIngressModel, key string, parse
...
-		}
-		evhNode.ServiceEngineGroup = lib.GetSEGName()
-		evhNode.VrfContext = lib.GetVrf()
-		evhNode.ApplicationProfile = utils.DEFAULT_L7_APP_PROFILE
-		// Remove the redirect for secure to insecure transition
-		hosts := []string{host}
-		if pathsvcmap.gslbHostHeader != "" {
-			hosts = append(hosts, pathsvcmap.gslbHostHeader)
+func (o *AviObjectGraph) BuildModelGraphForInsecureEVH(routeIgrObj RouteIngressModel, host, infraSettingName, key string, pathsvcmap HostMetada) {
+	vsNode := o.GetAviEvhVS()
```
Comment (akshayhavile) [inline internal/nodes/avi_model_evh_nodes.go:756]: We are missing lock in insecure model creation.

File: internal/nodes/avi_model_evh_nodes.go [hunk 3]
```diff
@@ -740,77 +740,83 @@ func ProcessInsecureHostsForEVH(routeIgrObj RouteIngressModel, key string, parse
...
+				NamespaceIngressName: ingressHostMap.GetIngressesForHostName(host),
+				Namespace:            namespace,
+				HostNames:            hostSlice,
+			},
 		}
-		RemoveRedirectHTTPPolicyInModelForEvh(evhNode, hosts, key)
 
-		// build poolgroup and pool
-		aviModel.(*AviObjectGraph).BuildPolicyPGPoolsForEVH(vsNode, evhNode, namespace, ingName, key, infraSettingName, hosts, pathsvcmap.ingressHPSvc)
 		foundEvhModel := FindAndReplaceEvhInModel(evhNode, vsNode, key)
```
Comment (akshayhavile) [inline internal/nodes/avi_model_evh_nodes.go:791]: This is not related to your changes. But FindAndReplaceEvhInModel and following if condition (line #791-794 ) seems redundant... as we are doing same check at line number #813.

File: internal/nodes/avi_model_evh_nodes.go (hunk 2 above)
Comment (monotosh-avi) [reply internal/nodes/avi_model_evh_nodes.go:756]: Done

File: internal/nodes/avi_model_evh_nodes.go (hunk 3 above)
Comment (monotosh-avi) [reply internal/nodes/avi_model_evh_nodes.go:791]: removed.

**Comment by monotosh-avi** [comment]
build ako

**Review by akshayhavile — APPROVED** [review]

**Comment by monotosh-avi** [comment]
build ako

**Review by chauhanshubham — APPROVED** [review]
